File: scraper.js
/**
 * FKF News RSS Scraper
 * Fetches https://footballkenya.org/news (React SPA → needs Puppeteer),
 * extracts article titles, URLs and dates, then writes fkf-news.xml.
 *
 * Usage:  node scraper.js
 * Cron:   0 * * * * node /path/to/scraper.js   (every hour)
 */

const puppeteer = require("puppeteer");
const RSS = require("rss");
const fs = require("fs");
const path = require("path");

const TARGET_URL = "https://footballkenya.org/news";
const OUTPUT_FILE = path.join(__dirname, "fkf-news.xml");
const MAX_ARTICLES = 20;

async function scrape() {
  console.log(`[${new Date().toISOString()}] Launching browser…`);

  const browser = await puppeteer.launch({
    headless: "new",
    args: ["--no-sandbox", "--disable-setuid-sandbox"],
  });

  const page = await browser.newPage();
  await page.setUserAgent(
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
      "(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
  );

  try {
    console.log(`Navigating to ${TARGET_URL}…`);
    await page.goto(TARGET_URL, { waitUntil: "networkidle2", timeout: 30000 });

    // Wait for at least one news card / article link to appear
    await page
      .waitForSelector("a[href*='/news/']", { timeout: 15000 })
      .catch(() => console.warn("Selector timeout – trying anyway…"));

    // Dump the rendered HTML for inspection if needed
    // fs.writeFileSync("debug.html", await page.content());

    const articles = await page.evaluate((max) => {
      const results = [];
      const seen = new Set();

      /**
       * Convert a URL slug to a readable title.
       * e.g. "coach-cheche-names-provisional-squad-as-junior-starlets-eye-uganda"
       *   → "Coach Cheche Names Provisional Squad As Junior Starlets Eye Uganda"
       */
      function slugToTitle(url) {
        try {
          const slug = new URL(url).pathname
            .split("/")
            .filter(Boolean)
            .pop() || "";
          return slug
            .replace(/-/g, " ")
            .replace(/\b\w/g, (c) => c.toUpperCase());
        } catch {
          return null;
        }
      }

      /**
       * Resolve a title from the DOM element hierarchy.
       * Prefers explicit headings, then aria-label, then slug.
       */
      function resolveTitle(a, parent, url) {
        const heading = parent?.querySelector("h1,h2,h3,h4,h5");
        const headingText = heading?.textContent?.trim();
        // Skip generic site-name headings
        if (headingText && headingText.toLowerCase() !== "football kenya federation") {
          return headingText;
        }
        const ariaLabel = a.getAttribute("aria-label")?.trim();
        if (ariaLabel && ariaLabel.toLowerCase() !== "football kenya federation") {
          return ariaLabel;
        }
        const aText = a.textContent?.trim();
        if (aText && aText.toLowerCase() !== "football kenya federation" && aText.length > 10) {
          return aText;
        }
        // Fall back to slug-derived title
        return slugToTitle(url) || "FKF News";
      }

      /**
       * Pick the best image from a parent container.
       * Skips the logo (fkf-logo) and favicon-style images.
       */
      function resolveImage(parent) {
        if (!parent) return null;
        const imgs = Array.from(parent.querySelectorAll("img"));
        const article = imgs.find(
          (img) =>
            img.src &&
            !img.src.includes("fkf-logo") &&
            !img.src.includes("favicon") &&
            img.naturalWidth > 80
        );
        return article?.src || imgs[0]?.src || null;
      }

      // Strategy 1: explicit <article> tags
      document.querySelectorAll("article").forEach((el) => {
        const a = el.querySelector("a[href]");
        const timeEl = el.querySelector("time");
        if (!a) return;
        const url = a.href;
        if (seen.has(url) || !url.includes("/news/")) return;
        seen.add(url);
        results.push({
          title: resolveTitle(a, el, url),
          url,
          date: timeEl?.getAttribute("datetime") || timeEl?.textContent?.trim() || null,
          image: resolveImage(el),
        });
      });

      // Strategy 2: anchor[href*='/news/'] card layout
      document.querySelectorAll("a[href*='/news/']").forEach((a) => {
        const url = a.href;
        if (seen.has(url)) return;
        seen.add(url);

        // Walk up max 7 levels to find a card container
        let parent = a.parentElement;
        for (let i = 0; i < 6; i++) {
          if (!parent) break;
          parent = parent.parentElement;
        }

        const timeEl = parent?.querySelector("time");

        results.push({
          title: resolveTitle(a, parent, url),
          url,
          date: timeEl?.getAttribute("datetime") || timeEl?.textContent?.trim() || null,
          image: resolveImage(parent),
        });
      });

      return results.slice(0, max);
    }, MAX_ARTICLES);

    console.log(`Extracted ${articles.length} articles.`);
    if (articles.length === 0) {
      console.warn(
        "No articles found. The page structure may have changed. " +
          "Run with headless:false to debug."
      );
    }

    // Build RSS feed
    const feed = new RSS({
      title: "Football Kenya Federation – Latest News",
      description:
        "Auto-generated RSS feed of the latest news from footballkenya.org",
      feed_url: "https://yourdomain.com/rss/fkf-news.xml", // ← update this
      site_url: "https://footballkenya.org",
      image_url: "https://footballkenya.org/favicon.png",
      language: "en",
      ttl: 60, // cache for 60 minutes
      pubDate: new Date(),
    });

    for (const article of articles) {
      feed.item({
        title: article.title,
        url: article.url,
        date: article.date ? new Date(article.date) : new Date(),
        description: article.image
          ? `<img src="${article.image}" alt="${article.title}" />`
          : article.title,
        enclosure: article.image
          ? { url: article.image, type: "image/jpeg" }
          : undefined,
      });
    }

    const xml = feed.xml({ indent: true });
    fs.writeFileSync(OUTPUT_FILE, xml, "utf8");
    console.log(`✅ RSS feed written to ${OUTPUT_FILE}`);
    console.log(
      `   Contains ${articles.length} items. Upload this file to your static host.`
    );
  } finally {
    await browser.close();
  }
}

scrape().catch((err) => {
  console.error("Scraper failed:", err.message);
  process.exit(1);
});

File: scraper.py
import os
import requests
from bs4 import BeautifulSoup
from email.utils import formatdate
import datetime
from xml.sax.saxutils import escape

def fetch_and_generate_rss(url, output_file):
    # 1. Fetch the web page
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
    }
    
    print(f"Fetching {url}...")
    try:
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
    except requests.exceptions.RequestException as e:
        print(f"Error fetching URL: {e}")
        return

    # 2. Parse HTML using BeautifulSoup
    soup = BeautifulSoup(response.content, 'html.parser')
    
    # Try to find article containers. Often they are <article> tags, or divs with class 'post' or 'article'
    articles = soup.find_all('article')
    if not articles:
        # Fallback: look for generic post wrappers
        articles = soup.find_all('div', class_=lambda c: c and ('post' in c or 'news-item' in c))

    rss_items = []
    
    for idx, article in enumerate(articles):
        # Extract URL and Title from standard heading tags or anchor tags
        # We look for an <a> tag inside a heading (h2, h3, h4) or the first prominent <a> tag
        heading = article.find(['h1', 'h2', 'h3', 'h4'])
        if heading and heading.find('a'):
            a_tag = heading.find('a')
        else:
            a_tag = article.find('a')
            
        if not a_tag or not a_tag.get('href'):
            continue
            
        title = a_tag.get_text(strip=True)
        link = a_tag['href']
        
        # Ensure the link is absolute
        if link.startswith('/'):
            link = "https://footballkenya.org" + link

        # Extract Date
        # Typically wrapped in <time> tag or a div/span with class containing 'date'
        date_str = None
        time_tag = article.find('time')
        if time_tag and time_tag.get('datetime'):
            date_str = time_tag['datetime']
        else:
            date_elem = article.find(class_=lambda c: c and 'date' in c)
            if date_elem:
                date_str = date_elem.get_text(strip=True)
        
        # Parse date to RSS standard format (RFC 822)
        # If we can't parse it, we default to the current time for the feed validity
        pub_date = formatdate(localtime=True)
        if date_str:
            try:
                # Try parsing standard ISO format if it's from a datetime attribute
                dt = datetime.datetime.fromisoformat(date_str.replace('Z', '+00:00'))
                pub_date = formatdate(dt.timestamp(), localtime=False)
            except ValueError:
                # If parsing fails, stick to current time for standard compliance
                pass

        rss_items.append({
            'title': title,
            'link': link,
            'pub_date': pub_date,
            'description': f"News article: {title}" # Could be enhanced to extract excerpts
        })
        
        if len(rss_items) >= 15: # Limit to latest 15 items
            break

    print(f"Extracted {len(rss_items)} articles.")

    # 3. Generate RSS XML
    # Using simple string formatting for the RSS structure
    rss_template = f"""<?xml version="1.0" encoding="UTF-8" ?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
<channel>
    <title>Football Kenya Federation News</title>
    <link>https://www.footballkenya.org/news</link>
    <description>Latest news from Football Kenya Federation, scraped dynamically.</description>
    <lastBuildDate>{formatdate(localtime=True)}</lastBuildDate>
"""
    
    for item in rss_items:
        rss_template += f"""
    <item>
        <title>{escape(item['title'])}</title>
        <link>{escape(item['link'])}</link>
        <pubDate>{item['pub_date']}</pubDate>
        <description>{escape(item['description'])}</description>
        <guid>{escape(item['link'])}</guid>
    </item>"""
        
    rss_template += """
</channel>
</rss>
"""

    # 4. Save to file
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(rss_template)
    
    print(f"Successfully wrote RSS feed to {output_file}")


if __name__ == "__main__":
    target_url = "https://footballkenya.org/news"
    # Or "https://footballkenya.org/category/news/" depending on actual URL structure
    
    # Write to a public folder or current directory
    output_xml_path = "fkf-news.xml"
    fetch_and_generate_rss(target_url, output_xml_path)
